### PR TITLE
[tt-train] Composite Muon implementation

### PR DESCRIPTION
**Pull request by zbaczewskiTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/38916

### Problem description
Training with Muon optimizer requires Newton-Schulz orthogonalization of weight updates, a composite Muon optimizer implementation, and Python-level integration to route parameters between Muon (2D weight matrices) and AdamW (embeddings, norms, LM head).

#### Newton-Schulz composite implementation
+ **`tt-train/sources/ttml/ops/newton_schulz_op.cpp`** — composite Newton-Schulz iteration (`newtonschulz5` with hardcoded coefficients, `newtonschulz` with configurable a/b/c)
+ **`tt-train/sources/ttml/ops/newton_schulz_op.hpp`** — public API
+ **`tt-train/tests/ops/newton_schulz_op_test.cpp`** — tests validating orthogonalization properties and convergence

#### MuonComposite optimizer (C++)
+ **`tt-train/sources/ttml/optimizers/muon.hpp`** — `MuonComposite` optimizer class with `MuonConfig` (lr, momentum, ns_steps)
+ **`tt-train/sources/ttml/optimizers/muon.cpp`** — momentum accumulation + Newton-Schulz orthogonalization step
+ **`tt-train/tests/optimizers/muon_test.cpp`** — accuracy validation against CPU float reference implementation
+ **`tt-train/sources/ttml/optimizers/optimizer_registry.cpp`** — registered `MuonComposite` in C++ optimizer registry
+ **`tt-train/sources/ttml/nanobind/nb_optimizers.cpp`** — Python bindings for `MuonComposite` and `MuonConfig`

#### MuonWithAdamW Python optimizer
+ **`tt-train/sources/ttml/ttml/common/muon_optimizer.py`** — `MuonWithAdamW` composite optimizer registered in Python optimizer registry. Routes 2D weight matrices to MuonComposite, everything else to AdamW. Configurable `adamw_modules` list for excluding specific modules (embeddings, LM head, fused KV linear)
+ **`tt-train/sources/examples/nano_gpt/train_nanogpt.py`** — registers `MuonWithAdamW` optimizer for use via YAML config

#### Configs
+ **`tt-train/configs/training_configs/training_shakespeare_tinyllama_muon.yaml`** — TinyLlama 1.1B training config with MuonWithAdamW

#### Other
+ **`tt-train/sources/ttml/CMakeLists.txt`** — added new source files to build
+ **`tt-train/tests/CMakeLists.txt`** — added new test files
+ **`tt-train/sources/ttml/ttnn_fixed/matmuls.cpp`** / **`matmuls.hpp`** — added optional `output_tensor` parameter to `matmul` to allow buffer reuse in Newton-Schulz iterations

### TinyLlama 1.1B, Blackhole P100a
image

Both MuonWithAdamW and AdamW converge to similar final loss (~0.1 vs ~0.25) on Shakespeare, but this is likely due to TinyLlama 1.1B heavily overfitting on the small dataset. 
Further training runs on larger datasets are needed to evaluate Muon efficacy

---

As can be seen, Muon adds over 700 ms to a single training step. There are two main reasons. First, the computation itself: each Newton-Schulz iteration performs 3 matmuls, and with 5 iterations that's 15 matmuls per parameter. Second, dispatch latency: in TinyLlama 1.1B there are 110 matrices optimized by Muon, each requiring ~40 kernel dispatches per step, totaling over 4000 kernel dispatches for a single training step.
```
  ┌───────────────────┬─────────────┬─────────────┐
  │       Phase       │ [2048,2048] │ [2048,5632] │
  ├───────────────────┼─────────────┼─────────────┤
  │ Momentum          │ 226 us      │ 403 us      │
  ├───────────────────┼─────────────┼─────────────┤
  │ NS Norm           │ 288 us      │ 506 us      │
  ├───────────────────┼─────────────┼─────────────┤
  │ NS Transpose      │ 29 us       │ 25 us       │
  ├───────────────────┼─────────────┼─────────────┤
  │ NS Loop (5 iters) │ 7,178 us    │ 14,376 us   │
  ├───────────────────┼─────────────┼─────────────┤
  │ NS Transpose back │ 38 us       │ 26 us       │
  ├───────────────────┼─────────────┼─────────────┤
  │ NS Total          │ 7,534 us    │ 14,933 us   │
  ├───────────────────┼─────────────┼─────────────┤
  │ Param Update      │ 241 us      │ 409 us      │
  ├───────────────────┼─────────────┼─────────────┤
  │ Total             │ 8,041 us    │ 15,774 us   │
  └───────────────────┴─────────────┴─────────────┘
```
*This benchmark was performed with synchronizations in between*

### Checklist

- [x] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:zbaczewski/muon)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:zbaczewski/muon)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:zbaczewski/muon)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/ttml/common/muon_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        for name in self._muon_param_names:
+            print(f"    - {name}")
+        print(f"  AdamW parameters ({len(self._adamw_param_names)}):")
+        for name in self._adamw_param_names:
+            print(f"    - {name}")
+
+
+def register():
+    ttml.optimizers.register_optimizer(
+        "MuonWithAdamW", lambda config, params: MuonWithAdamW(config, params)
```
[the remark this answers is not in the document]
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ttml/common/muon_optimizer.py]: Valid, fixed.

File: tt-train/sources/ttml/ttml/common/muon_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    def get_state_dict(self):
+        return {
+            "muon": self._muon.get_state_dict(),
+            "adamw": self._adamw.get_state_dict(),
+        }
+
+    def set_state_dict(self, dict):
+        self._muon.set_state_dict(dict["muon"])
+        self._adamw.set_state_dict(dict["adamw"])
```
[the remark this answers is not in the document]
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ttml/common/muon_optimizer.py]: valid, fixed

File: tt-train/tests/ops/newton_schulz_op_test.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    auto G_tensor = core::from_xtensor(G_data, &autograd::ctx().get_device());
+    auto X_tensor = ops::newtonschulz(G_tensor, 10, 1e-7f, a, b, c);
+    auto X_result = core::to_xtensor(X_tensor);
+
+    EXPECT_TRUE(xt::allclose(X_result, G_expected, 5e-2f, 5e-2f));
+
+    auto X_2d = xt::view(X_result, 0, 0, xt::all(), xt::all());
+    auto XXT = xt::linalg::dot(X_2d, xt::transpose(X_2d));
+    auto I = xt::eye<float>(32);
+    auto ortho_error = xt::abs(XXT - I);
```
[the remark this answers is not in the document]
Comment (zbaczewskiTT) [reply tt-train/tests/ops/newton_schulz_op_test.cpp]: valid, fixed

File: tt-train/tests/optimizers/muon_test.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        w_cpu = muon_step_cpu(w_cpu, g0, mom_cpu, tc.lr, tc.momentum, tc.ns_steps, i);
+    }
+
+    // Device
+    auto param = autograd::create_tensor(core::from_xtensor(w0, &autograd::ctx().get_device()), true);
+    param->set_grad(core::from_xtensor(g0, &autograd::ctx().get_device()));
+
+    serialization::NamedParameters params{{"theta", param}};
+    optimizers::MuonConfig config{.lr = tc.lr, .momentum = tc.momentum, .ns_steps = tc.ns_steps};
+    optimizers::Muon optimizer(params, config);
```
[the remark this answers is not in the document]
Comment (zbaczewskiTT) [reply tt-train/tests/optimizers/muon_test.cpp]: valid, fixed

File: tt-train/sources/ttml/optimizers/muon.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
+// SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "muon.hpp"
+
+#include <fmt/format.h>
+
+#include "autograd/auto_context.hpp"
+#include "autograd/autocast_tensor.hpp"
```
[the remark this answers is not in the document]
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/muon.cpp]: valid, fixed

**Review by vmelnykovTT — APPROVED** [review]

File: tt-train/sources/ttml/ops/newton_schulz_op.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+tt::tt_metal::Tensor newtonschulz(const tt::tt_metal::Tensor& G, int steps, float eps, float a, float b, float c) {
+    ttnn::Tensor X = G;
+
+    ttnn::Tensor squares = ttnn::square(X);
+    ttnn::Tensor sum_squares =
+        ttnn::sum(squares, ttsl::SmallVector<int>{-2, -1}, true, std::nullopt, core::ComputeKernelConfig::precise());
+    ttnn::Tensor norm_tensor = ttnn::sqrt(sum_squares);
+
+    ttnn::Tensor norm_plus_eps = ttnn::add(norm_tensor, eps);
+    X = ttnn::divide(X, norm_plus_eps);
```
Comment (philei-tt) [inline tt-train/sources/ttml/ops/newton_schulz_op.cpp:23]: Can't this be replaced with rmsnorm ops?

File: tt-train/sources/ttml/ops/newton_schulz_op.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        ttnn::addalpha(buf_A, buf_A2, c, std::nullopt, buf_A);
+
+        // X_new = a*X + B @ X: compute B @ X, then fuse add with alpha
+        ttnn_fixed::matmul(buf_A, X, false, false, buf_BX);
+        ttnn::addalpha(buf_BX, X, a, std::nullopt, X);
+    }
+
+    buf_A.deallocate();
+    buf_A2.deallocate();
+    buf_BX.deallocate();
```
Comment (philei-tt) [inline tt-train/sources/ttml/ops/newton_schulz_op.cpp]: Unnecessary, deallocate called on scope exit

File: tt-train/sources/ttml/optimizers/muon_composite.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        const auto& tensor_ptr = m_parameters.at(name);
+        if (!tensor_ptr->is_grad_initialized()) {
+            continue;
+        }
+
+        const auto gradients = tensor_ptr->get_grad();
+
+        if (m_steps > 0 && m_config.momentum != 0.0F) {
+            buffer = ttnn::multiply(buffer, m_config.momentum);
+            buffer = ttnn::add(buffer, gradients);
```
Comment (philei-tt) [inline tt-train/sources/ttml/optimizers/muon_composite.cpp:52]: try ttnn.addcmul here, might be better perf.

File: tt-train/sources/ttml/ttnn_fixed/matmuls.cpp [hunk 9]
```diff
@@ -8,7 +8,11 @@
 
 namespace ttml::ttnn_fixed {
 tt::tt_metal::Tensor matmul(
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttnn_fixed/matmuls.cpp:10]: Should you update python bindings too to add support for preallocated output tensor?

File: tt-train/tests/optimizers/muon_composite_test.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    auto w_device = core::to_xtensor(param->get_value());
+    EXPECT_TRUE(xt::allclose(w_device, w_cpu, 1e-2f, 1e-2f));
+}
+
+static const MuonTestCase kMuonCases[] = {
+    {{1, 1, 32, 128}, 1e-2f, 0.95f, 5, 1, "Wide"},
+    {{1, 1, 128, 32}, 1e-2f, 0.95f, 5, 2, "Tall_2_step"},
+    {{1, 1, 128, 128}, 1e-2f, 0.95f, 5, 3, "Square_3_step"},
+};
```
Comment (philei-tt) [inline tt-train/tests/optimizers/muon_composite_test.cpp:127]: Maybe add test cases with non-one batch and head dimension?

**Review by philei-tt — APPROVED** [review]

File: tt-train/sources/ttml/optimizers/muon_composite.cpp (hunk 8 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/optimizers/muon_composite.cpp:52]: From what I see addcmul performs `input+value×tensor1×tensor2` so it's not usable here

File: tt-train/sources/ttml/ops/newton_schulz_op.cpp (hunk 7 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ops/newton_schulz_op.cpp]: done

File: tt-train/tests/optimizers/muon_composite_test.cpp (hunk 10 above)
Comment (zbaczewskiTT) [reply tt-train/tests/optimizers/muon_composite_test.cpp:127]: We always optimize matrices; Newton-Schulz is defined for 2D matrices. Polar decomposition for stacked parameters flattened into 2D wouldn't make sense.

File: tt-train/sources/ttml/ttnn_fixed/matmuls.cpp (hunk 9 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ttnn_fixed/matmuls.cpp:10]: There are no bindings for ttnn_fixed::matmul

File: tt-train/sources/ttml/ops/newton_schulz_op.cpp (hunk 6 above)
Comment (zbaczewskiTT) [reply tt-train/sources/ttml/ops/newton_schulz_op.cpp:23]: frobenius norm computes norm of the whole flattened matrix
rmsnorm is computed and applied per-row
